pages/k8s/backups.md

```diff
@@ -0,0 +1,111 @@
+---
+
+# Backups
+
+The state of your Kubernetes cluster is kept in the etcd datastore. This page
+shows how to backup and restore the etcd included in the **Canonical
+Distribution of Kubernetes <sup>&reg;</sup>**.
+
+ Backing up application specific data, normally stored in a persistent volume, is not
+ currently supported by native Kubernetes. Various third party solutions are available -
+ please refer to their own documentation for details.
+
+ ## Creating an **etcd** snapshot
+
+ <div class="p-notification--warning"><p class="p-notification__response">
+ <span class="p-notification__status">Warning:</span>
+ Snapshots can only be restored on the **same version of etcd**.
+  </p></div>
+
+
+
+ **etcd** is a distributed key/value store. To create a snapshot, all that is required is to run
+ the `snapshot` action on one of the units running **etcd**:
+
+ ```bash
+ juju run-action etcd/0 snapshot --wait
+ ```
+
+ By specifying `--wait`, the console will wait to return the result of running
+ the action, which in this case includes the path and filename of the generated
+ snapshot file. For example:
+
+ ```bash
+ unit-etcd-0:
+  id: 3d6a505e-07d7-4697-8471-60156f87b1b4
+  results:
+    copy:
+      cmd: juju scp etcd/0:/home/ubuntu/etcd-snapshots/etcd-snapshot-2018-09-26-18.04.02.tar.gz
+        .
+    snapshot:
+      path: /home/ubuntu/etcd-snapshots/etcd-snapshot-2018-09-26-18.04.02.tar.gz
+      sha256: e85ae4d49b6a889de2063031379ab320cc8f09b6e328cdff2fb9179fc641eee9
+      size: 68K
+      version: |-
+        etcdctl version: 3.2.10
+        API version: 2
+  status: completed
+  timing:
+    completed: 2018-09-26 18:04:04 +0000 UTC
+    enqueued: 2018-09-26 18:04:04 +0000 UTC
+    started: 2018-09-26 18:04:03 +0000 UTC
+  unit: etcd/0
+  ```
+
+  This path/filename relates to the unit where the action was run. As we will
+  likely want to use the snapshot on a different unit, we should fetch the
+  snapshot to the local machine. The command to perform this is also helpfully
+  supplied in the `copy` section of the output (see above):
+
+```bash
+  juju scp etcd/0:/home/ubuntu/etcd-snapshots/etcd-snapshot-2018-09-26-18.04.02.tar.gz .
+  ```
+
+It is also wise to check the sha256 checksum of the file you have copied
+against the value in the previous output:
+
+```bash
+sha256sum etcd-snapshot-2018-09-26-18.04.02.tar.gz
+```
+
+## Restoring a snapshot
+
+<div class="p-notification--warning"><p class="p-notification__response">
+<span class="p-notification__status">Warning:</span>
+Restoring a snapshot should not be performed when there is more than one unit of **etcd** running.
+ </p></div>
+
+ As restoring only works when there is a single unit of **etcd**, it is usual to deploy a new
+instance of the application first. In this case, we can attache the snapshot as a resource during creation
+
+```bash
+juju deploy etcd new-etcd --series=xenial --resource snapshot=./etcd-snapshot-2018-09-26-18.04.02.tar.gz
```

Comment:
Here is what worked for me:
```
> cat ./newer-etcd.yaml
newer-etcd:
  channel: 3.2/stable
> juju deploy etcd newer-etcd --series=xenial --config=newer-etcd.yaml
Located charm "cs:etcd-126".
Deploying charm "cs:etcd-126".
> juju add-relation newer-etcd easyrsa
> juju attach newer-etcd snapshot=./etcd-snapshot-2018-10-04-12.12.32.tar.gz
> juju run-action newer-etcd/0 restore
Action queued with id: bbe7b0db-0222-4d25-89fb-1f717b6bcd00
> juju show-action-output bbe7b0db-0222-4d25-89fb-1f717b6bcd00
status: completed
timing:
  completed: 2018-10-04 12:37:15 +0000 UTC
  enqueued: 2018-10-04 12:36:49 +0000 UTC
  started: 2018-10-04 12:37:13 +0000 UTC
```

Reply:
many thanks for the review, especially this part, very helpful.

I will play around with the versions. I'm reluctant to 'hardwire' a specific version into the instructions if it can be avoided, though we should certainly show how to select a specific version

Reply:
This still doesn't work for me:


juju run-action etcd/0 snapshot --wait
juju scp etcd/0:/home/ubuntu/etcd-snapshots/etcd-snapshot-2018-10-25-12.40.02.tar.gz .
sha256sum etcd-snapshot-2018-10-25-12.40.02.tar.gz 

juju deploy etcd newer-etcd --series=bionic --config channel=3.2/stable
juju relate newer-etcd easyrsa
 juju attach newer-etcd snapshot=./etcd-snapshot-2018-10-25-12.40.02.tar.gz
 juju run-action newer-etcd/0 restore
juju show-action-output 0bf6bbc5-4866-48ff-8c4d-3e81d9581aee

            message: exit status 1
            status: failed
            timing:
              completed: 2018-10-25 12:53:54 +0000 UTC
              enqueued: 2018-10-25 12:53:41 +0000 UTC
              started: 2018-10-25 12:53:41 +0000 UTC



juju status "*etcd"

            Model                         Controller          Cloud/Region   Version  SLA          Timestamp
            conjure-canonical-kubern-7d4  conjure-up-aws-089  aws/eu-west-1  2.4.5    unsupported  13:57:48+01:00

            App         Version  Status  Scale  Charm  Store       Rev  OS      Notes
            etcd        3.2.10   active      3  etcd   jujucharms  209  ubuntu  
            newer-etcd  3.2.10   active      1  etcd   jujucharms  209  ubuntu  

            Unit           Workload  Agent  Machine  Public address  Ports     Message
            etcd/0*        active    idle   2        52.214.110.47   2379/tcp  Healthy with 3 known peers
            etcd/1         active    idle   3        34.247.47.176   2379/tcp  Healthy with 3 known peers
            etcd/2         active    idle   4        176.34.145.218  2379/tcp  Healthy with 3 known peers
            newer-etcd/0*  active    idle   15       18.203.250.35   2379/tcp  Healthy with 1 known peer

Reply:
This is interesting -- @ktsakalozos had etcd-126, while @evilnick had -209.  Looks like we need to verify restore on the newer etcd revs.

Reply:
@evilnick your script worked for me with both xenial and bionic versions of `newer-etcd`.  The only difference is that i specified v3 in the snapshot invocation:

    juju run-action etcd/0 snapshot keys-version=v3 --wait

Is it possible you did not?  My output, fyi:
```bash
$ juju scp etcd/0:/home/ubuntu/etcd-snapshots/etcd-snapshot-2018-10-29-15.00.37.tar.gz .
$ juju deploy etcd newer-etcd --series=bionic --config channel=3.2/stable
Located charm "cs:etcd-209".
Deploying charm "cs:etcd-209".
$ juju relate newer-etcd easyrsa
$ juju attach newer-etcd snapshot=./etcd-snapshot-2018-10-29-15.00.37.tar.gz
$ juju run-action newer-etcd/0 restore --wait
unit-newer-etcd-0:
  id: fd76aa8d-6635-429c-8313-0f42002f2f0b
  status: completed
  timing:
    completed: 2018-10-29 15:11:23 +0000 UTC
    enqueued: 2018-10-29 15:11:19 +0000 UTC
    started: 2018-10-29 15:11:21 +0000 UTC
  unit: newer-etcd/0
```

Reply:
I'll try running through it again. 
So, on version 3.x it defaults to snapshotting in v2 format, which it then can't restore?

Reply:
> So, on version 3.x it defaults to snapshotting in v2 format, which it then can't restore?

@evilnick, indeed, I think that is what's happening.  I've opened https://github.com/juju-solutions/layer-etcd/issues/143 to choose a more appropriate value vs defaulting `v2` in the action.  For this doc, I think it's worth being explicit with the snapshot `keys-version` param so users know when they might have to tailor that to their deployment.